File: src/searching/cpp/interpolation_search/src/interpolation_search.cpp
#include "interpolation_search.h"

#include <algorithm>
#include <cstddef>
#include <optional>
#include <vector>

std::optional<std::size_t> interpolation_search(const std::vector<int> &arr, int target) {
  if (arr.empty()) {
    return std::nullopt;
  }

  std::size_t low = 0;
  std::size_t high = arr.size() - 1;

  while (low <= high && target >= arr[low] && target <= arr[high]) {
    // Handle case where all elements are the same
    if (arr[high] == arr[low]) {
      if (arr[low] == target) {
        return low;
      }
      return std::nullopt;
    }

    // Estimate position using interpolation formula
    std::size_t pos = low + static_cast<std::size_t>(
        (static_cast<double>(target - arr[low]) / (arr[high] - arr[low])) * (high - low)
    );

    // Ensure pos is within bounds
    pos = std::max(low, std::min(pos, high));

    if (arr[pos] == target) {
      return pos;
    } else if (arr[pos] < target) {
      low = pos + 1;
    } else {
      if (pos == 0) {
        break;
      }
      high = pos - 1;
    }
  }

  return std::nullopt;
}

File: src/searching/python/interpolation_search/src/interpolation_search.py
"""Interpolation search algorithm implementation."""

from typing import List, Optional


def interpolation_search(array: List[int], target: int) -> Optional[int]:
    """
    Search for a target value in a sorted array using interpolation search.

    This algorithm estimates the likely position of the target using the values
    themselves, then probes there and narrows the search based on comparison.
    Works best on uniformly distributed sorted data.

    Time complexity: O(log log n) average for uniform data, O(n) worst case
    Space complexity: O(1)

    Args:
        array: The sorted list of integers to search.
        target: The value to search for.

    Returns:
        The index of the target if found, None otherwise.

    Note:
        The array must be sorted in ascending order for correct results.
        Works best when values are uniformly distributed.
    """
    low = 0
    high = len(array) - 1

    while low <= high and target >= array[low] and target <= array[high]:
        # Handle case where all elements are the same
        if array[high] == array[low]:
            if array[low] == target:
                return low
            return None

        # Estimate position using interpolation formula
        pos = low + ((target - array[low]) * (high - low)) // (array[high] - array[low])

        # Ensure pos is within bounds
        pos = max(low, min(pos, high))

        if array[pos] == target:
            return pos
        elif array[pos] < target:
            low = pos + 1
        else:
            high = pos - 1

    return None
